ls1intum/Artemis
# [Feature] add clone & download repo button in programming details view

### Checklist

### Motivation and Context


Resolves #2552 

It should be possible for instructors to clone and/or download a zipped version of the test, template, and solution repositories (separately) for programming exercises. 

### Description

The programming exercise details view displays the URLs for the template, test, and solution repositories. This PR adds a "clone repo" and "download repo" button next to each URL.

This PR does the following:
-  Extracts the existing clone repo button into its own `jhi-clone-repo-button` component so that it can be used in views other than the student view
-  Creates a new component `jhi-programming-exercise-instructor-repo-download` which enables instructors to download the repo as a zipped file
- Adds a new endpoint `/export-instructor-repository/{repositoryTypeString}` that enabled instructors to export a test, template, or solution repository 

### Steps for Testing


1. Log in to Artemis
2. Navigate to Course Administration -> Exercise
3. Click on a programming exercise and navigate to where the URLs are located
4. You should see the buttons
5. Test that the zipped files are the same of the respective repository
6. The "Clone Repo" functionality should behave the same way as it does for students

### Test Coverage



programming-exercise-detail.component.ts 50.47%
programming-exercise.service.ts 61.9%
programming-exercise-instructor-repo-download.component.ts 100%
exercise-details-student-actions.component.ts 77.05%
clone-repo-button.component.ts 94.44%
ProgrammingExerciseExportService.java 48.1% 
ProgrammingExerciseResource.java   87.1%
GitService.java  74.5%

### Screenshots


![image](https://user-images.githubusercontent.com/11298674/105505659-82593b80-5cc9-11eb-98d3-c601accdeb19.png)
![image](https://user-images.githubusercontent.com/11298674/105505713-956c0b80-5cc9-11eb-89a4-1050c5ca00c2.png)
